File: hw8/LinkedList.java


@SuppressWarnings("unchecked")
public class LinkedList<E> implements Queue<E> , List<E>{
	
	public LinkedList(){
		size = 0 ;
		capacity = 1 ;
		arr = (E[])new Object[capacity];
	}

	public iterator<E> iterator(){
		it = new iterator<E>(arr);
		return it;
	}
	public boolean add(E element){
		if(size >= capacity){
			resize();
		}
		arr[size] = element ;
		size++;
		return true ;
	}
	public boolean addAll(Collection<E> elements){
		iterator<E> it = elements.iterator();
		while(it.hasNext()){
			add(it.next());
		}
		return true ;		
	}
	public void clear(){
		arr = (E[])new Object[0];
		size = 0 ; 
		capacity = 1 ;
	}
	public boolean contains(E element){
		for(int i = 0 ; i < size() ; i++){
			if (element == arr[i]){
				return true ;
			}			
		}
		return false ;
	}
	public boolean containsAll(Collection<E> elements){
		int found = 0;
		
		if (elements.size() > size()){
			return false ;
		}
		for(int i =0 ;i < size() ; i++){
			iterator<E> it = elements.iterator();
			while(it.hasNext()){
				if(it.next() == get(i)){
					found++;
				}
			}
		}
		if (found >= elements.size()){
			return true ;
		}
		
		return false ;
	}
	public boolean isEmpty(){
		if(size == 0 ){
			return true ;
		}
		return false ;
	}
	public void remove(E element){
		int index = indexOf(element);
		E[] temp = arr; 
		if ( index != -1){
			capacity--;
			arr = (E[])new Object[capacity];
			for (int i = 0 ,j=0; i < size() ; i++){
				if( i != index ){
					arr[j] = temp[i];
					j++;
				}
			}
		}
		size = capacity;
	}
	public void removeAll(Collection<E> elements){
		iterator<E> it = elements.iterator();
		while(it.hasNext()){
			remove(it.next());
		}
			
	}
	public void retainAll(Collection<E> elements){
		boolean found  ;
		for(int i = 0 ; i < size() ; i++){
			iterator<E> it = elements.iterator();
			found = false;
			while(it.hasNext()){
				if(it.next() == get(i)){
					found = true ; 
				}
			}
			if(!found){
				remove(get(i));
				i--;
			}
		}
	}
	public int size(){
		return size ;
	}
	public E element(){
		try{
			if(!isEmpty()){
				return arr[0];
			}
			else {
				throw new IllegalAccessException("empty queue \n");
			}
		}catch(IllegalAccessException e){
			System.out.printf(e.getMessage());
		}
		return null;
	}
	public boolean offer(E element){
		try{
			if(size < capacity){
				add(element);
				return true ;
			}
			else {
				throw new IllegalAccessException("no space in the queue \n");
			}
		}catch(IllegalAccessException e){
			System.out.printf(e.getMessage());
		}
		return false;
	}
	public E poll(){
		if (!isEmpty()){
			E temp = arr[0];
			remove(arr[0]);
			return temp ;
		}
		return null;
	}
	public Object[] toArray(){
		Object[] temp = new Object[size()];
		for(int i = 0 ; i < size() ; i++){
			temp[i] = arr[i];
		}
		return temp ;
	}
	public boolean equals(Object o){
		if (o instanceof LinkedList){
			LinkedList<E> a = (LinkedList<E>)o;
			if (a.size() != size()){
				return false ;
			}
			for(int i = 0 ; i < a.size() ; i++){
				if(a.get(i) != get(i) ){
					return false ;
				}
			}
		}
		return true ;
	}
	public int indexOf(Object o){
		for(int i = 0 ; i < size() ; i++){
			if ((E)o == get(i)){
				return i ;
			}
		}
		return -1 ;
	}

	public E get(int index){
		try{
			if(index < size){
				return arr[index];
			}
			else {
				throw new IllegalAccessException("out of range \n");
			}
		}catch(IllegalAccessException e){
			System.out.printf(e.getMessage());
		}
		return null;
	}

	public String toString(){
		return String.format("LinkedList of length ( %d )\n",size);
	}

	private void resize(){
		E[] temp = arr ;
		capacity += 1 ;
		arr = (E[])new Object[capacity];
		for (int i =0 ; i < size ; i++ ){
			arr[i] = temp [i];
		}
	}

	private E[] arr ;
	private int size ;
	private int capacity;
	private iterator<E> it ;
}

File: hw6/LinkedList.cpp
# include "LinkedList.h"
namespace jwan_ {

	template<class E, template <typename... > class Container>
	Iterator<E,Container> LinkedList<E,Container>::iterator(){

		return Iterator<E,Container>(container);
	}

	template<class E, template <typename... > class Container>
	int LinkedList<E,Container>::size(){
		return container.size() ;
	}

	template<class E, template <typename... > class Container>
	bool LinkedList<E,Container>::isEmpty(){
		return container.empty();
	}

	template<class E, template <typename... > class Container>
	E LinkedList<E,Container>::operator [] (int index){
		Iterator<E,Container> temp ;
		temp = this->iterator();

		if (index < container.size() && index >= 0 ){
			for(int i = 0 ; i <= index ; i++){
				temp.next();
			}
			return *temp; 
		}
		std::cout<<"out of range !! ";
		
	}

	template<class E, template <typename... > class Container>
	bool LinkedList<E,Container>::contains(E e){
		Iterator<E,Container> temp ;
		temp = this->iterator();

		while(temp.hasnext()){
			if ( *temp == e ){
				return true ;
			}
			temp.next();
		}
		return false ;
	}

	template<class E, template <typename... > class Container>
	bool LinkedList<E,Container>::containsAll(Collection<E,Container>& c){
		
		LinkedList<E,Container>* v = dynamic_cast< LinkedList<E,Container>* > (&c);
		if (this->size() < v->size()){
			return false ;
		}
		Iterator<E,Container> temp ;
		temp = this->iterator();
		int counter =0 ;
		while (temp.hasnext()){
			Iterator<E,Container> temp1 = v->iterator();
			while(temp1.hasnext()){
				if (*temp == *temp1){
					counter++;
				}
				temp1.next();
			}
			temp.next();
		}
		if (counter >= v->size()){
			return true ;
		}
		return false ;
	}

	template<class E, template <typename... > class Container>
	void LinkedList<E,Container>::insert(E e){
		container.insert(container.end() , e );
		
	}

	template<class E, template <typename... > class Container>
	void LinkedList<E,Container>::clear(){
		container.clear();		
	}

	template<class E, template <typename... > class Container>
	void LinkedList<E,Container>::remove(E e) {
		auto temp = container.begin();
		
		for (int i = 0 ; i < container.size() ; i++){
			if ( *temp == e ){
				container.erase(temp);
			}
			temp++;
		}		
	}

	template<class E, template <typename... > class Container>
	void LinkedList<E,Container>::removeAll(Collection<E,Container>& c){
		LinkedList<E,Container>* v = dynamic_cast< LinkedList<E,Container>* > (&c);	
		auto temp = v->iterator();
		for (int i = 0 ; i < v->size() ; i++){
			this->remove(*temp);
			temp.next();
		}
	}

	template<class E, template <typename... > class Container>
	void LinkedList<E,Container>::add(E e){
		this->insert(e);
	}

	template<class E, template <typename... > class Container>
	void LinkedList<E,Container>::addAll(Collection<E,Container>& c){
		LinkedList<E,Container>* v = dynamic_cast< LinkedList<E,Container>* > (&c);	
		auto temp = v->iterator();
		for (int i =0 ; i < v->size() ; i++){
			this->add(*temp);
			temp.next();
		}
	}

	template<class E, template <typename... > class Container>
	void LinkedList<E,Container>::retainAll(Collection<E,Container>& c){
		LinkedList<E,Container>* v = dynamic_cast< LinkedList<E,Container>* > (&c);	
		auto temp = this->iterator();
		int found=0; 
		while (temp.hasnext()){
			found = 0 ;
			auto temp1 = v->iterator();
			while(temp1.hasnext()){
				if ( *temp == *temp1 ){
					found = 1 ;
				}
				temp1.next();
			}
			if(found == 0 ){
				this->remove(*temp);
			}
			temp.next();
		}
	}

	template<class E, template <typename... > class Container>
	void LinkedList<E,Container>::pop(){
		container.erase(container.begin());		
	}

	template<class E, template <typename... > class Container>
	Iterator<E,Container> LinkedList<E,Container>::element(){
		return this->iterator();
		
	}

	template<class E, template <typename... > class Container>
	void LinkedList<E,Container>::offer(E e){
		this->add(e);		
	}

	template<class E, template <typename... > class Container>
	Iterator<E,Container> LinkedList<E,Container>::poll(){
		container.erase(container.begin());
		return this->element();
	}


}